### ISSUE TITLE
Working with remote desktop software when using keyboard key selection source

### ISSUE THREAD

**Issue by annakirkpatrick** [OptiKey/OptiKey#772]
To start with, let me acknowledge that I think this issue really comes from a limitation/design decision in remote desktop software, not from a bug in Optikey. Nonetheless, if we can find a workaround, it would be good to document that, as I am sure I'm not the only one who has experienced or will experience this issue.

I use OptiKey as a main method of computer access. I use the keyboard key selection source for both key and point selection. (I actually use a set of external physical switches. These are connected to my computer via a USB switch interface, which sends the keypress F23 when a switch is pressed.)

I recently needed to use a remote desktop client. I tried using the Windows Remote Desktop connection tool. I was able to successfully fill out the connection details, enter my password, and actually connect to the remote host using Optikey. But once I got connected, I could no longer activate any keys on the Optikey keyboard.

I think I know what's happening. Whenever the Remote Desktop client is in focus, ALL keystrokes are sent to the remote host. This includes the F23 keystroke that I use to trigger selections in Optikey. Therefore, Optikey never sees the keystroke and can't respond to it.

I reached out to the Microsoft Disability Answers Desk, and they confirmed that the incompatibility between switch access systems and the Windows Remote Desktop connection tool is a known limitation.

When I switched to the dwell selection method, Optikey worked as expected. (Unfortunately, using the dwell selection method is not viable for me. It's much slower, and it gives me a headache.) This further confirms that the issue is with the keyboard key selection method specifically.

I suspect, therefore, that the real issue here is the design decisions behind the remote desktop software and therefore we can't fix this behavior by changing something in Optikey. But please tell me if you think I'm wrong.

Does anyone have any suggestions for possible workarounds? Perhaps other remote desktop software would be better? Or another approach that I'm not thinking of?

If we can pull together some solid information about workarounds (or even just a good understanding of exactly what the limitations are), I would be happy to write up some user documentation regarding OptiKey and Remote Desktop clients.

**Comment by annakirkpatrick** [comment]
>> I actually use a set of external physical switches. These are connected to my computer via a USB switch interface, which sends the keypress F23 when a switch is pressed.

> Do you give us more details about it?

The switches that I use to "click" on keys are standard assistive technology switches. E.g. Any of the products discussed on this page: https://enablingdevices.com/product-category/switches/

The switches are connected by an X keys switch interface: https://xkeys.com/xk3swi.html

The X keys switch interface is programmed so that, when a switch is pressed, the keystroke F23 is sent to the computer. (The computer "thinks" the switch interface is just a USB keyboard.)

>> I would be happy to write up some user documentation regarding OptiKey and Remote Desktop clients.

> Do you need to access machines others?

I'm not sure what you are asking here. 

I am trying to access other computers remotely. It's something I need to do for my job.

The only purpose of my comment was to offer to write some documentation for the wiki (or similar) so that it is easier for future users. (Of course, I have to solve the underlying technical problem first, before I can write documentation about it.)

**Comment by AdamRoden** [comment]
@annakirkpatrick It has been a long time since I've used X-keys, but I know they have a MacroWorks application that you can use to do more. Remote desktop software hijacks your mouse and keyboard hardware input, but if you run your X-keys in software mode, (rather than hardware mode,) I'm almost certain you can send the input to your local computer and application of choice.
https://xkeys.com/software/softwarewindows/softwaremacroworks.htm

**Comment by tqphan** [comment]
@annakirkpatrick I can't test this at the moment but I remember being able to prevent RDC from eating up the inputs by running it in a window instead of fullscreen.

**Comment by kmcnaught** [comment]
Hi @annakirkpatrick, hopefully one of the previous comments helps, but if not, I was curious and had a bit of a google and have a couple of possible ideas for investigation (you may of course have already tried some/all of them!)

1 - it's worth looking in the Autohotkey forums as that community may face similar issues and have some workarounds. If there's something that works with AHK but not Optikey then at least we have gained understanding or possibly might have something we can change in Optikey itself. For instance [this AHK forum thread](https://autohotkey.com/board/topic/5380-getting-keystrokes-thru-to-remote-desktop-client/) discusses different behaviour depending on fullscreen/window mode, and with different order of initialisation, though I confess I only skimmed it.. 

2 - it sounds like the Windows key shortcuts are 'special' with regard to how they are handled when using RDC. It might be a crazy idea but perhaps you could disable windows key shortcuts locally and then use the Windows key as your switch input, if your switch interface supports it in hardware? This wouldn't be a very nice workaround but if nothing else works it could be worth trying...

**Comment by annakirkpatrick** [comment]
@tqphan you are totally right! Thank you so much for the suggestion. Running the Remote Desktop client in a windowed mode does solve the problem. (To be honest, I didn't realize that there was a windowed mode until I went digging in the settings.)

Also thank you @AdamRoden and @kmcnaught for your suggestions! I really appreciate the community support.

As for documentation, there admittedly isn't much to say about this. But may be a wiki page under the troubleshooting section would be appropriate? I could easily write something up. It would be quite short.

**Comment by kmcnaught** [comment]
@annakirkpatrick @JuliusSweetland  I'm closing this as resolved (and also hopefully superseded by gamepad support), but it would be great to have more user-driven documentation for "how to get around common hurdles", perhaps a single high-level wiki page with links to standalone user guides? Please feel free to send over markdown files
